openvdb/openvdb/tools/Merge.h

```diff
@@ -0,0 +1,931 @@
+/// @details A union and an intersection are opposite operations to each other so
+/// implemented in a combined class. Use the CsgUnionOp and CsgIntersectionOp aliases
+/// for convenience.
+template<typename TreeT, bool Union>
+struct CsgUnionOrIntersectionOp
+{
+    using ValueT = typename TreeT::ValueType;
+    using RootT = typename TreeT::RootNodeType;
+    using LeafT = typename TreeT::LeafNodeType;
+
+    /// @brief Convenience constructor to CSG union or intersect a single
+    /// non-const tree with another. This constructor takes a Steal or DeepCopy
+    /// tag dispatch class.
+    template <typename TagT>
+    CsgUnionOrIntersectionOp(TreeT* tree, TagT tag) { mTreesToMerge.emplace_back(tree, tag); }
+
+    /// @brief Convenience constructor to CSG union or intersect a single
+    /// const tree with another. This constructor requires a DeepCopy tag
+    /// dispatch class.
+    CsgUnionOrIntersectionOp(const TreeT* tree, DeepCopy tag) { mTreesToMerge.emplace_back(tree, tag); }
+
+    /// @brief Constructor to CSG union or intersect a container of multiple
+    /// const or non-const tree pointers. A Steal tag requires a container of
+    /// non-const trees, a DeepCopy tag will accept either const or non-const
+    /// trees.
+    template <typename TreesT, typename TagT>
+    CsgUnionOrIntersectionOp(TreesT& trees, TagT tag)
+    {
+        for (auto& tree : trees) {
+            mTreesToMerge.emplace_back(tree, tag);
+        }
+    }
+
+    /// @brief Constructor to accept a vector of TreeToMerge objects, primarily
+    /// used when mixing const/non-const trees.
+    /// @note Union/intersection order is preserved.
+    explicit CsgUnionOrIntersectionOp(const std::vector<TreeToMerge<TreeT>>& trees)
+        : mTreesToMerge(trees) { }
+
+    /// @brief Constructor to accept a deque of TreeToMerge objects, primarily
+    /// used when mixing const/non-const trees.
+    /// @note Union/intersection order is preserved.
+    explicit CsgUnionOrIntersectionOp(const std::deque<TreeToMerge<TreeT>>& trees)
+        : mTreesToMerge(trees.cbegin(), trees.cend()) { }
+
+    /// @brief Return true if no trees being merged
+    bool empty() const { return mTreesToMerge.empty(); }
+
+    /// @brief Return the number of trees being merged
+    size_t size() const { return mTreesToMerge.size(); }
+
+    // Processes the root node. Required by the NodeManager
+    bool operator()(RootT& root, size_t idx) const;
+
+    // Processes the internal nodes. Required by the NodeManager
+    template<typename NodeT>
+    bool operator()(NodeT& node, size_t idx) const;
+
+    // Processes the leaf nodes. Required by the NodeManager
+    bool operator()(LeafT& leaf, size_t idx) const;
+
+private:
+    // on processing the root node, the background value is stored, retrieve it
+    // and check that the root node has already been processed
+    const ValueT& background() const;
+
+    mutable std::vector<TreeToMerge<TreeT>> mTreesToMerge;
```

Comment:
minor: it's weird that this has to be mutable when the container doesn't change. Maybe this should be a vector of pointers.

Reply:
Yeah, I don't love this aspect of this implementation. There's a lot I successfully managed to change about the NodeManager implementation, but the requirement for having a const operator when doing a foreach operation was a bridge too far. Right now, it's just a detail of the class, so it's something we could theoretically come back and change at a later date.